components/builder-jobsrv/src/server/worker_manager.rs

```diff
@@ -491,12 +527,18 @@ impl WorkerMgr {
                     };
                     worker.quarantine();
                 } else {
-                    worker.busy(job_id);
+                    worker.refresh();
                 }
             }
             (jobsrv::WorkerState::Busy, jobsrv::WorkerState::Ready) => {
-                self.delete_worker(&worker)?;
-                worker.ready()
+                if !self.is_job_complete(worker.job_id.unwrap())? {
+                    // Handle potential race condition where a Ready heartbeat
```

Comment:
I'm so sorry for this. It's because we're going through two sockets instead of one - I fixed all of this in the other servers but didn't refactor the network code for JobSrv->Worker

![tenor-51344680](https://user-images.githubusercontent.com/54036/31538432-694b65d8-afba-11e7-8005-ae98644bda3c.gif)